### PR TITLE
add design for TLS config

### PR DESCRIPTION
**Pull request by deads2k**
The first step is in https://github.com/openshift/api/pull/432

@openshift/sig-network-edge 
@openshift/sig-master

### CODE REVIEW

File: apiserver/tls-config.md [hunk 1]
```diff
@@ -0,0 +1,362 @@
+---
+title: TLS-Config
+authors:
+  - "@deads2k"
+reviewers:
```
Comment (ironcladlou) [inline apiserver/tls-config.md]: @danehans should be here

File: apiserver/tls-config.md [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+last-updated: 2019-09-17
+status: implementable
+see-also:
+replaces:
+superseded-by:
+---
+
+# TLS Config for Externally Facing Services
+
+Cluster-admins need to be able to choose which cipher suites their kube-apiserver, oauth-server, and routers use.
```
Comment (ironcladlou) [inline apiserver/tls-config.md]: s/routers/ingress controllers throughout

File: apiserver/tls-config.md [hunk 3]
```diff
@@ -0,0 +1,362 @@
...
+## Summary
+
+Cluster-admins need to be able to choose which cipher suites their kube-apiserver, oauth-server, and routers use.
+This is a balance of corporate policy and compatibility with existing systems.  This should be a shared, top level
+configuration amongst the affected components with a local override for routers to allow workloads to have a different 
+level of TLS config restrictions than the infrastructure.
+
+The idea is to create a set of standard profiles: old, intermediate, modern, and custom that can be used.  Old, intermediate, 
+and modern are intent based.  We will update them as we see fit to stay up to date with standards as time goes on.  Custom
+allows a user to specify exactly which options they way.
```
Comment (ironcladlou) [inline apiserver/tls-config.md]: s/way/want

File: apiserver/tls-config.md [hunk 4]
```diff
@@ -0,0 +1,362 @@
...
+### Goals
+
+1. Easily configure a uniform level of TLSConfig (ciphers, tls versions, and DH key size) across kube-apiserver, oauth-server,
+ and all routers. These are the only known exposed services today.
+2. Allow overriding on a per-router basis.  Workloads and infrastructure may have different requirements and we need to 
+ support that use-case.
+
+### Non-Goals
+
+1. Make unusual configurations easy to make work.
```
Comment (ironcladlou) [inline apiserver/tls-config.md]: another way: "Make it easy to implement unusual configurations"

File: apiserver/tls-config.md [hunk 5]
```diff
@@ -0,0 +1,362 @@
...
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift/docs]
+
+## Summary
+
+Cluster-admins need to be able to choose which cipher suites their kube-apiserver, oauth-server, and routers use.
+This is a balance of corporate policy and compatibility with existing systems.  This should be a shared, top level
+configuration amongst the affected components with a local override for routers to allow workloads to have a different 
+level of TLS config restrictions than the infrastructure.
```
Comment (danehans) [inline apiserver/tls-config.md]: Is it necessary to restate the point from the TLS Config for Externally Facing Services section here?

File: apiserver/tls-config.md [hunk 6]
```diff
@@ -0,0 +1,362 @@
...
+
+The idea is to create a set of standard profiles: old, intermediate, modern, and custom that can be used.  Old, intermediate, 
+and modern are intent based.  We will update them as we see fit to stay up to date with standards as time goes on.  Custom
```
Comment (danehans) [inline apiserver/tls-config.md]: Can you elaborate on how w'll keep this static list updated? Are there any repercussions if the list is updated and we do not implement the updates in a timely manner? The project maintains a json list of recommendations [here](https://statics.tls.security.mozilla.org/server-side-tls-conf-5.0.json). We may want to programmatically manage our profiles based on this list.

File: apiserver/tls-config.md [hunk 7]
```diff
@@ -0,0 +1,362 @@
...
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift/docs]
+
+## Summary
+
+Cluster-admins need to be able to choose which cipher suites their kube-apiserver, oauth-server, and routers use.
```
Comment (danehans) [inline apiserver/tls-config.md]: Will the enhancement cover other tls config aspects (i.e. tls curves) in the future or be specific to cipher suites?

File: apiserver/tls-config.md [hunk 8]
```diff
@@ -0,0 +1,362 @@
...
+		Ciphers: []string{
+		},
+		TLSVersion: TLSVersion{
+			MaximumVersion: VersionTLS12,
+			MinimumVersion: VersionTLS12,
+		},
+		DHParamSize: DHParamSize2048,
+	},
+	TLSProfileModernType: {
+		Ciphers: []string{
```
Comment (danehans) [inline apiserver/tls-config.md]: Are the `Ciphers` empty on purpose for each profile?

Comment (deads2k) [reply apiserver/tls-config.md]: > Are the `Ciphers` empty on purpose for each profile?

yeah, too long for a KEP. This just shows the intent.

File: apiserver/tls-config.md (hunk 6 above)
Comment (deads2k) [reply apiserver/tls-config.md]: > Can you elaborate on how w'll keep this static list updated? Are there any repercussions if the list is updated and we do not implement the updates in a timely manner? The project maintains a json list of recommendations [here](https://statics.tls.security.mozilla.org/server-side-tls-conf-5.0.json). We may want to programmatically manage our profiles based on this list.

see below in proposal for the details of using the `TLSProfiles` map.

File: apiserver/tls-config.md (hunk 7 above)
Comment (deads2k) [reply apiserver/tls-config.md]: > Will the enhancement cover other tls config aspects (i.e. tls curves) in the future or be specific to cipher suites?

If it fits and is needed, we can consider it in a future enhancement, but not in this one.

Comment (deads2k) [reply apiserver/tls-config.md]: adding to non-goals

**Comment by deads2k** [comment]
comments addressed in separate commit.

File: kube-apiserver/tls-config.md [hunk 9]
```diff
@@ -0,0 +1,366 @@
...
+
+// TLSProfileSpec is the desired behavior of a TLSSecurityProfile.
+type TLSProfileSpec struct {
+	// ciphers is used to specify the cipher algorithms that are negotiated
+	// during the TLS handshake.  For example, to use 3DES  (yaml):
+	//
+	//   ciphers:
+	//     - 3DES
+	//
+	Ciphers []string `json:"ciphers"`
```
Comment (enj) [inline kube-apiserver/tls-config.md]: Are we doing no validation on this?  i.e. we expect the operators to all tolerate random junk here?  I am mostly trying to understand what happens if you manage to reduce the whitelist to something bad and then you cannot easily connect back to the kube API to undo the change.

File: kube-apiserver/tls-config.md [hunk 10]
```diff
@@ -0,0 +1,366 @@
...
+## Motivation
+
+We want guardrails where possible to avoid letting a customer make choices that prevent future upgrades.  Lots of flexibility
+is very easy to produce once, but very hard to maintain in an auto-updating world.  We will use profiles to constrain that choice.
+
+### Goals
+
+1. Easily configure a uniform level of TLSConfig (ciphers, tls versions, and DH key size) across kube-apiserver, oauth-server,
+ and all ingress controllers. These are the only known exposed services today.
+2. Allow overriding on a per-router basis.  Workloads and infrastructure may have different requirements and we need to
```
Comment (enj) [inline kube-apiserver/tls-config.md]: ingress controller

File: kube-apiserver/tls-config.md (hunk 9 above)
Comment (danehans) [reply kube-apiserver/tls-config.md]: > Are we doing no validation on this? i.e. we expect the operators to all tolerate random junk here?

@enj will a validating admission webhook suffice?

File: kube-apiserver/tls-config.md [hunk 11]
```diff
@@ -0,0 +1,366 @@
...
+	// modern is a TLS security profile based on:
+	//
+	// https://wiki.mozilla.org/Security/Server_Side_TLS#Modern_compatibility
+	//
+	// +optional
+	// +nullable
+	Modern *ModernTLSProfile `json:"modern,omitempty"`
+	// custom is a user-defined TLS security profile. Be extremely careful using a custom
+	// profile as invalid configurations can be catastrophic. An example custom profile
+	// looks like this:
```
Comment (danehans) [inline kube-apiserver/tls-config.md]: no example is provided, so remove `An example custom profile
	// looks like this:`.

File: kube-apiserver/tls-config.md (hunk 9 above)
Comment (enj) [reply kube-apiserver/tls-config.md]: Sure but we will have to relax its validation over time which can have weird consequences on downgrade.  Also I do not think we can assume that all components can honor all ciphers.

File: kube-apiserver/tls-config.md [hunk 12]
```diff
@@ -0,0 +1,366 @@
...
+// NOTE: The caller needs to make sure to check that these constants are valid for their binary. Not all
+// entries map to values for all binaries.  In the case of ties, the kube-apiserver wins.  Do not fail,
+// just be sure to whitelist only and everything will be ok.
+// TODO write the specifics of the ciphers, that's just too long to place in this KEP, it is in the openshift/api pull.
+var TLSProfiles = map[TLSProfileType]*TLSProfileSpec{
+	TLSProfileOldType: {
+		Ciphers: []string{
+		},
+		TLSVersion: TLSVersion{
+			MaximumVersion: VersionTLS10,
```
Comment (danehans) [inline kube-apiserver/tls-config.md]: I believe you have the values for `MaximumVersion` and `MinimumVersion` reversed. According to https://wiki.mozilla.org/Security/Server_Side_TLS#Old_backward_compatibility, `TLSProfileOldType` should support TLS 1.0-1.3

File: kube-apiserver/tls-config.md [hunk 13]
```diff
@@ -0,0 +1,366 @@
...
+			MinimumVersion: VersionTLS12,
+		},
+		DHParamSize: DHParamSize1024,
+	},
+	TLSProfileIntermediateType: {
+		Ciphers: []string{
+		},
+		TLSVersion: TLSVersion{
+			MaximumVersion: VersionTLS12,
```
Comment (danehans) [inline kube-apiserver/tls-config.md]: According to https://wiki.mozilla.org/Security/Server_Side_TLS#Intermediate_compatibility_.28recommended.29, `MaximumVersion ` should support `VersionTLS13`.

File: kube-apiserver/tls-config.md [hunk 14]
```diff
@@ -0,0 +1,366 @@
...
+			MinimumVersion: VersionTLS12,
+		},
+		DHParamSize: DHParamSize2048,
+	},
+	TLSProfileModernType: {
+		Ciphers: []string{
+		},
+		TLSVersion: TLSVersion{
+			MaximumVersion: VersionTLS12,
```
Comment (danehans) [inline kube-apiserver/tls-config.md]: According to https://wiki.mozilla.org/Security/Server_Side_TLS#Modern_compatibility, `MaximumVersion ` and `MinimumVersion ` should be `VersionTLS13`.

File: kube-apiserver/tls-config.md [hunk 15]
```diff
@@ -0,0 +1,366 @@
...
+			MinimumVersion: VersionTLS12,
+		},
+		DHParamSize: DHParamSize2048,
```
Comment (danehans) [inline kube-apiserver/tls-config.md]: According to https://wiki.mozilla.org/Security/Server_Side_TLS#Modern_compatibility `DHParamSize` is nil.

**Comment by deads2k** [comment]
@danehans we aren't going to keep this enhancement up to date with what's missing versus present.

**Comment by danehans** [comment]
> @danehans we aren't going to keep this enhancement up to date with what's missing versus present.

@deads2k ack. https://github.com/openshift/enhancements/pull/21/files#r327279420 still needs to be addressed as it conflicts with with https://github.com/openshift/enhancements/pull/21/files#diff-a2436a7ae013617df3999525dcb376c1R187-R206

**Comment by deads2k** [comment]
final comments addressed.  https://github.com/openshift/api/pull/432 updated.  @danehans lgty and apply the fine tooth comb to the API?
